src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,143 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
+            numiter >= maxiter && break
+            keep = max(floor(Int, (3 * krylovdim + 2 * converged) / (5 * bs)), 1) * bs
+            H = zeros(S, keep + bs, keep)
+            # The last bs rows of U contribute to calculate errors of Ritz values.
+            @inbounds for j in 1:keep
+                H[j, j] = D[j]
+                H[(keep + 1):end, j] = U[(K - bs + 1):K, j]
+            end
+            # Turn diagonal matrix D into a block tridiagonal matrix, and make sure 
+            # The residual of krylov subspace keeps the form of [0,..,0,R]
+            @inbounds for j in keep:-1:1
+                h, ν = householder(H, j + bs, 1:j, j)
+                H[j + bs, j] = ν
+                H[j + bs, 1:(j - 1)] .= zero(eltype(H))
+                lmul!(h, H)
+                rmul!(view(H, 1:(j + bs - 1), :), h')
+                rmul!(U, h')
+            end
+            # Transform the basis and update the residual and update the BTD.
+            BTD .= S(0)
+            BTD[1:keep, 1:keep] .= H[1:keep, 1:keep]
+            B = basis(fact)
+            basistransform!(B, view(U, :, 1:keep))
+
+            r_new = OrthonormalBasis(fact.r.vec[1:bs_r])
+            view_U = view(U, (K - bs_r + 1):K, (keep - bs_r + 1):keep)
+            basistransform!(r_new, view_U)
+            fact.r.vec[1:bs_r] = r_new[1:bs_r]
+
+            while length(fact) > keep
+                pop!(fact.V)
+                fact.k -= 1
+            end
+            numiter += 1
+        end
+    end
+
+    howmany_actual = howmany
+    if converged > howmany
+        howmany_actual = converged
+    elseif length(D) < howmany
+        howmany_actual = length(D)
+    end
+    U1 = view(U, :, 1:howmany_actual)
+    vectors = let V = basis(fact)
+        [V * u for u in cols(U1)]
+    end
+    bs_r = fact.r_size
+    K = length(fact)
+    U2 = view(U, (K - bs_r + 1):K, 1:howmany_actual)
+    R = fact.r
+    residuals = [zerovector(x₀[1]) for _ in 1:howmany_actual]
+    @inbounds for i in 1:howmany_actual
+        for j in 1:bs_r
+            residuals[i] = add!!(residuals[i], R[j], U2[j, i])
+        end
+    end
+    normresiduals = normresiduals[1:howmany_actual]
+
+    if (converged < howmany) && verbosity >= WARN_LEVEL
+        @warn """BlockLanczos eigsolve stopped without full convergence after $(K) iterations:
+        * $converged eigenvalues converged
+        * norm of residuals = $(normres2string(normresiduals))
+        * number of operations = $numops"""
+    elseif verbosity >= STARTSTOP_LEVEL
+        @info """BlockLanczos eigsolve finished after $(K) iterations:
+        * $converged eigenvalues converged
+        * norm of residuals = $(normres2string(normresiduals))
+        * number of operations = $numops"""
+    end
+
+    return D[1:howmany_actual],
+           vectors[1:howmany_actual],
+           ConvergenceInfo(converged, residuals, normresiduals, numiter, numops)
```

Comment:
Here, the slicing can now be dropped by using `values`, and `vectors` is already of the correct length:

```suggestion
    return values, vectors,
           ConvergenceInfo(converged, residuals, normresiduals, numiter, numops)
```

Reply:
I will remove the slicing. However, to stay consistent with the formatting in the Lanczos code, I will keep `values` and `vectors` on separate lines:
```julia
    return values,
           vectors,
           ConvergenceInfo(converged, residuals, normresiduals, numiter, numops)
```